apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/FlushableConsumerBase.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// Base class for the <see cref="IFlushableConsumer{T}"/> that exposes a <see cref="SemaphoreSlim"/> to
+    /// protect the buffer.
+    /// </summary>
+    internal abstract class FlushableConsumerBase<T> : IFlushableConsumer<T>
+    {
+        /// <summary>A semaphore to protect the buffer.</summary>
+        protected readonly SemaphoreSlim Semaphore = new SemaphoreSlim(1);
```

Comment:
I would try to hide this if possible, perhaps by defining protected helper methods (one for sync and one for async logic), which do something like this:

```csharp
protected void HoldSemaphoreAndDo(Action action)
{
    _semaphore.Wait();
    try
    {
        action();
    }
    finally
    {
        _semaphore.Release();
    }
}
```

And then invoking these with lambdas. Although I guess you need to weigh that against the cost of allocating closure objects for each invocation. Perhaps it might be ok to just leave the semaphore exposed since the class is internal.

However, it might be good to remove the /// comment, and add a // comment saying something like "Note: if the class is made public, hide this and expose helpers to perform operators with the semaphore held." That way, if in the future the class is made public, there will be a warning about /// missing and then the author will hopefully see the warning and note and not unintentionally expose this.